### PR TITLE
feat: error codes

### PR DESCRIPTION
**Pull request by lizaiv77**
- introduce config with error codes 
- refactor deserialization so that it would have access to context with error codes
- additional type of exception that can be unwrapped into error code

### CODE REVIEW

File: utam-compiler/src/main/java/utam/compiler/grammar/ErrorProducer.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+import utam.core.framework.consumer.UtamError;
+
+/**
+ * This exception is thrown from deserializer when we do not know the context. Then we catch it and
+ * process
+ *
+ * @author elizaveta.ivanova
+ * @since 238
+ */
+class ErrorProducer extends RuntimeException {
```
Comment (jimevanssfdc) [inline utam-compiler/src/main/java/utam/compiler/grammar/ErrorProducer.java:24]: If we're going to be `throw`ing these objects, wouldn't it make sense for the class to be named something ending in `Exception` or `Error`?

File: utam-core/src/main/java/utam/core/declarative/translator/ErrorMessageProvider.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package utam.core.declarative.translator;
+
+/**
+ * API responsible for building compiler error messages
+ *
+ * @author elizaveta.ivanova
+ * @since 238
+ */
+public interface ErrorMessageProvider {
```
Comment (jimevanssfdc) [inline utam-core/src/main/java/utam/core/declarative/translator/ErrorMessageProvider.java:16]: Are there other interfaces we anticipate will be `ErrorMessageProvider`s? If not, why break out the separate interface?

File: utam-compiler/src/main/java/utam/compiler/types/BasicElementInterface.java [hunk 3]
```diff
@@ -135,7 +135,7 @@ public static boolean isBasicType(TypeProvider type) {
       return values.toArray(String[]::new);
     }
     if(isThrowAnError) {
-      throw new UtamCompilationError(typeNodeValueError);
+      throw new IllegalArgumentException(typeNodeValueError);
```
Comment (jimevanssfdc) [inline utam-compiler/src/main/java/utam/compiler/types/BasicElementInterface.java:138]: Why the exception type change? Not saying it's the wrong thing to do, just curious.

File: utam-compiler/src/main/java/utam/compiler/helpers/TranslationContext.java [hunk 4]
```diff
@@ -246,7 +240,7 @@ public ElementContext getElement(String name) {
       return DOCUMENT_ELEMENT;
     }
     if (!elementContextMap.containsKey(name)) {
-      throw new UtamError(String.format(ERR_CONTEXT_ELEMENT_NOT_FOUND, name));
+      throw new NullPointerException();
```
Comment (jimevanssfdc) [inline utam-compiler/src/main/java/utam/compiler/helpers/TranslationContext.java:243]: Is `NullPointerException` the correct choice here? Nothing is really "null," just that the key wasn't found in the map.

File: utam-compiler/src/main/java/utam/compiler/helpers/TranslationContext.java [hunk 5]
```diff
@@ -37,24 +38,16 @@
...
- * instance is created for every Page Object that is being translated <br> contains reference to the
- * Page Object context and translator context
+ * Instance of this type is created for every Page Object that is being translated, contains
+ * reference to the Page Object context and translator context
  *
  * @author elizaveta.ivanova
  * @since 228
  */
-public final class TranslationContext {
+public final class TranslationContext  {
```
Comment (jimevanssfdc) [inline utam-compiler/src/main/java/utam/compiler/helpers/TranslationContext.java:47]: Extra space?

File: utam-compiler/src/main/resources/errors.config.json [hunk 6]
```diff
@@ -0,0 +1,59 @@
+{
+  "P001": {
```
Comment (lizaiv77) [inline utam-compiler/src/main/resources/errors.config.json:2]: codes stand for:
- Pxxx means it's general parsing error that can happen for any entity
- UA stands for UtamArgument error
- UE for UtamElement error etc.
Abbreviation is there to easier navigate through codes

**Review by jimevanssfdc — CHANGES_REQUESTED** [review]
Have you compiled `utam-java-framework` with the compiler with this PR applied to make sure there's nothing that was `public` that no longer is, or that has differing behavior from what is expected by the Maven plugins in that repo?

File: utam-compiler/src/main/java/utam/compiler/grammar/ErrorProducer.java (hunk 1 above)
Comment (lizaiv77) [reply utam-compiler/src/main/java/utam/compiler/grammar/ErrorProducer.java:24]: renamed to UtamCompilerErrorWithCode

File: utam-compiler/src/main/java/utam/compiler/helpers/TranslationContext.java (hunk 5 above)
Comment (lizaiv77) [reply utam-compiler/src/main/java/utam/compiler/helpers/TranslationContext.java:47]: removed

File: utam-compiler/src/main/java/utam/compiler/helpers/TranslationContext.java (hunk 4 above)
Comment (lizaiv77) [reply utam-compiler/src/main/java/utam/compiler/helpers/TranslationContext.java:243]: addressed by adding error code as a parameter and throwing compiler error

File: utam-compiler/src/main/java/utam/compiler/types/BasicElementInterface.java (hunk 3 above)
Comment (lizaiv77) [reply utam-compiler/src/main/java/utam/compiler/types/BasicElementInterface.java:138]: refactored to throw compiler error

File: utam-core/src/main/java/utam/core/declarative/translator/ErrorMessageProvider.java (hunk 2 above)
Comment (lizaiv77) [reply utam-core/src/main/java/utam/core/declarative/translator/ErrorMessageProvider.java:16]: removed

File: utam-compiler/src/main/java/utam/compiler/UtamCompilerIntermittentError.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+import utam.compiler.helpers.TranslationContext;
+
+/**
+ * This exception is thrown from deserializer when we do not know the context. Then we catch it and
+ * process
+ *
+ * @author elizaveta.ivanova
+ * @since 238
+ */
+public class UtamCompilerIntermittentError extends RuntimeException {
```
Comment (jimevanssfdc) [inline utam-compiler/src/main/java/utam/compiler/UtamCompilerIntermittentError.java:23]: I'd call this `Intermediate` or `Interstitial` as opposed to `Intermittent`. The latter seems to imply "it happens sometimes."

**Review by jimevanssfdc — CHANGES_REQUESTED** [review]
One minor class name change, and we should be good to go.

**Review by jimevanssfdc — APPROVED** [review]

**Pull request merged**